### PR TITLE
Fix generated column evaluation based on default constraint.

### PR DESCRIPTION
**Pull request by seut**
When evaluating generated columns, possible default constraints of referenced columns must be taken into account.

### CODE REVIEW

File: common/src/main/java/io/crate/common/collections/Maps.java [hunk 1]
```diff
@@ -92,9 +93,23 @@ public static Object getByPath(Map<String, Object> value, List<String> path) {
...
+    }
+
+    /**
+     * Inserts a value into source under the given key+path
+     */
+    public static void mergeInto(Map<String, Object> source,
+                                 String key,
+                                 List<String> path,
+                                 Object value,
+                                 TriConsumer<Map<String, Object>, String, Object> writer) {
```
Comment (mfussenegger) [inline common/src/main/java/io/crate/common/collections/Maps.java]: Using a interface from `log4j` for this feels a bit weird. Would it be worth creating a interface in our own package?

File: sql/src/main/java/io/crate/execution/dml/upsert/GeneratedColsFromRawInsertSource.java [hunk 2]
```diff
@@ -78,4 +92,24 @@ public BytesReference generateSource(Object[] values) throws IOException {
         return BytesReference.bytes(XContentFactory.jsonBuilder().map(source));
     }
 
+    private Map<Reference, Object> buildDefaults(List<Reference> defaults,
+                                                 EvaluatingNormalizer normalizer,
+                                                 TransactionContext txnCtx) {
+        HashMap<Reference, Object> m = new HashMap<>();
+        for (Reference ref : defaults) {
+            Symbol defaultExpression = normalizer.normalize(ref.defaultExpression(), txnCtx);
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/execution/dml/upsert/GeneratedColsFromRawInsertSource.java]: Would it make sense to use `SymbolEvaluator` here instead, given that we want to evaluate it to a value? 

From its javadocs:

> This should be preferred over {@link EvaluatingNormalizer} if it's required to fully evaluate the tree

File: sql/src/main/java/io/crate/execution/dml/upsert/InsertSourceFromCells.java [hunk 3]
```diff
@@ -44,74 +47,118 @@
...
     InsertSourceFromCells(TransactionContext txnCtx,
                           Functions functions,
                           DocTableInfo table,
                           String indexName,
                           GeneratedColumns.Validation validation,
                           List<Reference> targets) {
-        this.targets = targets;
-        ReferencesFromInputRow referenceResolver = new ReferencesFromInputRow(targets, table.partitionedByColumns(), indexName);
+        // Ensure targets are mutable because defaults may be added
+        this.targets = new ArrayList<>(targets);
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/execution/dml/upsert/InsertSourceFromCells.java]: Is this safe for re-use?

In the `TransportShardUpsertAction` we create a single `InsertSourceFromCells` instance and then use it for multiple params. 

What if there is a null value early and later on there isn't?

File: sql/src/main/java/io/crate/execution/dml/upsert/InsertSourceFromCells.java [hunk 4]
```diff
@@ -44,74 +47,118 @@
...
+        ArrayList<Object> defaultValues = new ArrayList<>();
+        for (Reference ref : table.defaultExpressionColumns()) {
+            if (targets.contains(ref) == false) {
+                targets.add(ref);
+                Symbol defaultExpression = normalizer.normalize(ref.defaultExpression(), txnCtx);
+                if (defaultExpression instanceof Literal) {
+                    defaultValues.add(((Literal) defaultExpression).value());
+                } else {
+                    // this should never happen as default expression should always evaluate to a literal
+                    defaultValues.add(null);
```
Comment (mfussenegger) [inline sql/src/main/java/io/crate/execution/dml/upsert/InsertSourceFromCells.java]: Should we use the `SymbolEvaluator` here as well?

File: sql/src/main/java/io/crate/execution/dml/upsert/GeneratedColsFromRawInsertSource.java (hunk 2 above)
Comment (seut) [reply sql/src/main/java/io/crate/execution/dml/upsert/GeneratedColsFromRawInsertSource.java]: right, my bad.

File: common/src/main/java/io/crate/common/collections/Maps.java (hunk 1 above)
Comment (seut) [reply common/src/main/java/io/crate/common/collections/Maps.java]: yes wasn't sure if its worth, will do.

File: sql/src/main/java/io/crate/execution/dml/upsert/InsertSourceFromCells.java (hunk 3 above)
Comment (seut) [reply sql/src/main/java/io/crate/execution/dml/upsert/InsertSourceFromCells.java]: Should be safe yes, its only mutated once to add missing columns with a default constraint.  The default value of a target table column with a default constraint will be applied if the column is not explicitly defined for writing. Everything will be the same for all rows.

Example:
```
CREATE TABLE t1 (id int, x int);
CREATE TABLE t2 (id int, x int default 1, y as x + 1);
INSERT INTO t1 (id, x) VALUES (1, 3);
INSERT INTO t2 (id) (SELECT id, x FROM t1);         -> t2.x = 1 (default) for all rows
INSERT INTO t2 (id, x) (SELECT id, x FROM t1);      -> t2.x = t1.x for all rows
```

Comment (seut) [reply sql/src/main/java/io/crate/execution/dml/upsert/InsertSourceFromCells.java]: I've pushed a fixup which uses a slightly different pattern and also avoids array copies by using a new `BiArrayRow`.

File: dex/src/main/java/io/crate/data/BiArrayRow.java [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+ * software solely pursuant to the terms of the relevant commercial
+ * agreement.
+ */
+
+package io.crate.data;
+
+/**
+ * A {@link ArrayRow} variant which uses 2 backing arrays which can be changed.
+ */
+public class BiArrayRow implements Row {
```
Comment (mfussenegger) [inline dex/src/main/java/io/crate/data/BiArrayRow.java]: This is similar to `CombinedRow` (which requires two backing Row instances, which could be `RowN` or `ArrayRow` to accomplish the same), maybe that could be utilized instead.

Also fine for me to keep the extra instance if you prefer it that way.

Comment (seut) [reply dex/src/main/java/io/crate/data/BiArrayRow.java]: Yes right, this impl would safe the additional `Row` instances. No strong opinion here, shall I use `CombinedRow` instead?

Comment (mfussenegger) [reply dex/src/main/java/io/crate/data/BiArrayRow.java]: Also no strong opinion, both is fine I think.

**Comment by seut** [comment]
@mfussenegger I missed one scenario, pushed a fixup.

**Comment by jayeff** [comment]
Cool. Thanks for the fast turn-around! 👍

**Pull request merged**